Проект спринта 11.

KyleKiske opened the pull request:
Реализована работа с базой данных.
Переработаны классы, которые отвечали за хранение информации в памяти. Теперь данные хранятся в БД.
Добавлены эндпоинты для обращения к Жанрам и рейтингу MPA.

KyleKiske committed:
Реализована работа с базой данных.
Переработаны классы, которые отвечали за хранение информации в памяти. Теперь данные хранятся в БД.
Добавлены эндпоинты для обращения к Жанрам и рейтингу MPA.

KyleKiske committed:
В тест добавлена реализация встроенной для тестов базы данных, чтобы отделить её от основной.

KyleKiske committed:
Добавлены тесты для проверки эндпоинтов со встроенной БД.
Film и User заменены на Optional<Film> и Optional<User> в контроллерах соответственно.

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/controller/ErrorHandler.java:
```diff
@@ -1,30 +1,35 @@
-@RestControllerAdvice(basePackages = "ru.yandex.practicum.filmorate.controller")
+@RestControllerAdvice
+@Slf4j
 public class ErrorHandler {
```
Отличная обработка ошибок!

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -20,40 +18,40 @@ public FilmController(FilmService filmService){
 
     @GetMapping
-    public List<Film> findAll() {
+    public List<Optional<Film>> findAll() {
```
Возвращать Optional в контроллере неправильно. Пару причин. 
1. Если у вас не найдено ни одного элементе - стоит возвращать пустую коллекцию
2. Это последний слой работы приложения перед взаимодействием с клиентом, поэтому все объекты из контроллера будут сериализоваться, и тип Optional добавляет дополнительную обертку надо объектами, которые засоряют и усложняют сериализацию

Вот ещё статья на эту тему 
https://www.baeldung.com/java-optional-return#when-to-not-return-optional

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -21,27 +20,24 @@ public UserController(UserService userService){
 
     @PostMapping
-    public ResponseEntity<User> create(@RequestBody User user) {
+    public Optional<User> create(@RequestBody User user) {
```
При создании - у вас не может быть Optional, т.к. созданием контролируете вы, если у вас какие то ошибки - выкидывайте ошибку или возвращайте сущность. 
Optional нужно в 1 случае, когда мы ищем сущность по id, но даже в этом случае Optional нужно обрабатывать на слое БД, а дальше уже или выкидывать ошибку или работать с сущностью, если она найдена по id. 
Т.е. на слое контроллера в методе findById - будет возвращена сущность (не Optional) а в слое БД в методе findById() может быть Optional

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -21,27 +20,24 @@ public UserController(UserService userService){
 
     @GetMapping("/{id}")
-    public ResponseEntity<User> getUser(@PathVariable Long id){
+    public Optional<User> getUser(@PathVariable Long id){
```
Здесь нужно и во всех остальных контроллерах не должно быть Optional

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/dao/DBFilmRepository.java:
```diff
@@ -0,0 +1,176 @@
+    public Optional<Film> mapRowToFilm(ResultSet rs, int rowNum) throws SQLException {
+        int id = rs.getInt("FILM_ID");
+        return Optional.of(new Film(id, rs.getString("NAME"),
+                rs.getString("DESCRIPTION"),
```
Можно строки NAME, MPA и остальные, от которых зависит работа приложения - вынести в константу на уровень класса, хорошая практика

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/dao/DBFilmRepository.java:
```diff
@@ -0,0 +1,176 @@
+    @Override
+    public List<Optional<Film>> getFilms() {
+        final String sqlQuery = "SELECT FILM.FILM_ID, FILM.NAME, " +
+                "DESCRIPTION, RELEASE_DATE, DURATION, MPA, M.MPA_NAME FROM FILM " +
```
Когда используете alias - проставляйте его везде, когда обращаетесь к конретной таблице, а то у вас 
FILM.FILM_ID,
FILM.NAME,
DESCRIPTION (тут не ясно из какой таблицы ? из FILM или MPA ? )

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/dao/DBFilmRepository.java:
```diff
@@ -0,0 +1,176 @@
+        List<Optional<Film>> filmsOptional = jdbcTemplate.query(sqlQuery, this::mapRowToFilm, id);
+        List<Film> films = new ArrayList<>();
+        for (Optional<Film> f: filmsOptional){
+            films.add(f.get());
```
Здесь вы вызываете get(), хотя ещё не проверили, что Optional не пустой

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/dao/DBFilmRepository.java:
```diff
@@ -0,0 +1,176 @@
+        for (Optional<Film> f: filmsOptional){
+            films.add(f.get());
+        }
+        if(filmsOptional.size() == 0) {
```
Не верная проверка, нужно проверять optional.isPresent() 

+ если хотите проверить есть ли элементы в коллекции, проверяйте list.isEmpty() это best practies

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/dao/DBUserRepository.java:
```diff
@@ -0,0 +1,170 @@
+                        "PrimaryUserFriends.NAME, " +
+                        "PrimaryUserFriends.BIRTHDAY " +
+                        "FROM " +
+                            "(SELECT u.USER_ID, u.EMAIL, u.LOGIN, u.NAME, u.BIRTHDAY " +
```
У вас в одном запросе используется 2 под запроса, в целом всегда подзапросы плохо, т.к. на 1 запрос будет N+1 запрос, а здесь аж N+2, подумайте как обойтись без подзапроса

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,158 +1,76 @@
-                log.warn(exception.getMessage());
-                return new ResponseEntity<>(null, HttpStatus.NOT_FOUND);
-            }
+    public Optional<Film> getFilm(int id){
```
Вот здесь сделайте так
```
public Film getFilm(int id) {
        Optional<Film> film = filmRepository.getFilmById(id);
        if (film.isPresent()) {
            return film.get();
        } else {
            throw new FilmNotFoundException("Фильм с id " + id + " не найден");
        }
    }
```

или так:
```
public Film getFilm(int id) {
        return filmRepository.getFilmById(id)
                .orElseThrow(() -> new FilmNotFoundException("Film with id not found " + id));
    }
```

KyleKiske committed:
В контроллерах убраны все Optional значения.
Переделан запрос ан получение общих друзей.
В запросе на получение фильма по Id добавлены идентификаторы таблиц, для улучшения читаемости.
